### PR TITLE
Add header input type to support multi headers

### PR DESCRIPTION
**Pull request by NinaFan0824**
1. Some headers have default values, for example, the assignment rule header is default to true, which we want to update https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/headers_autoassign.htm 
2. We also need some headers to make Salesforce Conditional updates https://github.com/vacasaoss/ts-force/pull/5, at the meantime, handle the errors that is likely to be returned when the object doesn't match the conditions. https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/intro_rest_conditional_requests.htm

Currently, ts-force doesn't support adding headers. Based on this, we add some headers to send in patch requests. The below are the headers we intend to add.
```
Sforce-Auto-Assign: FALSE
If-Match: "Jbjuzw7dbhaEG3fd90kJbx6A0ow=-gzip", "U5iWijwWbQD18jeiXwsqxeGpZQk=-gzip"
If-None-Match: "Jbjuzw7dbhaEG3fd90kJbx6A0ow=-gzip", "U5iWijwWbQD18jeiXwsqxeGpZQk=-gzip"
If-Modified-Since: Tue, 10 Aug 2015 00:00:00 GMT
If-Unmodified-Since: Tue, 10 Aug 2015 00:00:00 GMT
Sforce-Mru: updateMru=true
x-sfdc-packageversion-clientPackage: 1.0 //x-sfdc-packageversion-[namespace]: xx.x , where [namespace] is the unique namespace of the managed package and xx.x is the package version.
```

### CODE REVIEW

File: ts-force/src/rest/restObject.ts [hunk 1]
```diff
@@ -229,27 +238,34 @@ export abstract class RestObject extends SObject {
...
+        GetOrHeadRequestHeaders.includes(key as keyof RequestHeadersInput)
+      );
+      const containsPatchOrPostHeader = Object.keys(opts.headers).some((key) =>
+        PatchOrPostRequestHeaders.includes(key as keyof RequestHeadersInput)
       );
+      if (containsGetOrHeadHeader && response.status === 304) {
+        throw new Error('Not Modified');
+      } else if (containsPatchOrPostHeader && response.status === 412) {
+        throw new Error('Precondition Failed');
+      }
```
Comment (jackholmes1010) [inline ts-force/src/rest/restObject.ts]: why is this required?

Comment (colinfindlay-nz) [reply ts-force/src/rest/restObject.ts]: Also - you shouldn't be throwing generic errors
https://github.com/vacasaoss/ts-force/blob/master/ts-force/src/rest/errors.ts

File: ts-force/src/rest/restHeader.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+export type RequestHeadersInput = {
+  'Sforce-Auto-Assign'?: boolean;
+  'Sforce-Mru'?: SforceMru;
+  'x-sfdc-packageversion'?: PackageVersion;
+  'If-Modified-Since'?: Date;
+  'If-Unmodified-Since'?: Date;
+  'If-Match'?: string[];
+  'If-None-Match'?: string[];
+  'Generic-Header'?: GenericHeader;
```
Comment (ChuckJonas) [inline ts-force/src/rest/restHeader.ts]: Is `Generic-Header` just allowing an arbitrary header key/value?  

If so, it seems like you would want to potentially support multiple values.  It might also be easier to just allow a `Record<string, string>`.

`additionalHeaders?: Record<string, string>;`

File: ts-force/src/rest/restObject.ts [hunk 3]
```diff
@@ -229,27 +228,29 @@ export abstract class RestObject extends SObject {
...
-        { headers }
+        { headers: headerOutput }
       );
+
+      // Check if it contains Conditional Request Headers and deal with errors
+      const containsConditionalHeader = Object.keys(opts.headers).some((key) =>
+        ConditionalRequestHeaders.includes(key as keyof RequestHeadersInput)
+      );
+      if (containsConditionalHeader && (response.status === 304 || response.status === 412)) {
+        throw new ConditionalError('Conditions not met');
```
Comment (ChuckJonas) [inline ts-force/src/rest/restObject.ts]: Might want to include the header keys in the exception

File: ts-force/src/rest/restHeader.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+};
+
+export const ConditionalRequestHeaders: (keyof RequestHeadersInput)[] = [
+  'If-Match',
+  'If-None-Match',
```
Comment (ChuckJonas) [inline ts-force/src/rest/restHeader.ts]: I've never actually used these `etag` caching headers with Salesforce... Reviewing [the docs](https://developer.salesforce.com/docs/atlas.en-us.api_rest.meta/api_rest/resources_sobject_retrieve_get.htm), I see this statement:

> An optional header specifying a comma-delimited list of one or more ETags. This only has an effect when used with Account objects. The request is only processed if the Account’s ETag does not match one of the ETags in the list.

Does it really only work with `Account`?

Comment (NinaFan0824) [reply ts-force/src/rest/restHeader.ts]: From the document [here](https://developer.salesforce.com/docs/atlas.en-us.uiapi.meta/uiapi/ui_api_get_started_intro.htm), Etag is used for http caching. I guess that’s the reason why it’s said only work with Account. I test with a a case project, and `If-Match` and `If-None-Match` can be added on get request but not on patch request.
Screenshot 2024-02-23 at 9 20 14 AM
Screenshot 2024-02-23 at 9 19 46 AM

File: ts-force/src/test/rest/restObjects.spec.ts [hunk 5]
```diff
@@ -6,8 +6,6 @@ import { getCalendarDate } from '../../utils/calendarDate';
 import { Account, Contact, User } from '../assets/sobs';
```
Comment (jackholmes1010) [inline ts-force/src/test/rest/restObjects.spec.ts]: I would recommend undoing the formatting changes

File: ts-force-gen/package.json [hunk 6]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@vacasaoss/ts-force-gen",
-  "version": "3.4.4",
+  "version": "3.4.5",
```
Comment (jackholmes1010) [inline ts-force-gen/package.json]: Is this a breaking change? Should we increment the major version?

**Review by jackholmes1010 — APPROVED** [review]
